Modernize the webserver code (part 2)

timvandermeij opened the pull request:
The commit messages contain more details about the individual changes. This second part completes the extraction of self-contained functionality from the large handler method to make it more manageable and to get rid of using variables from the outer scope, and enables the ESLint `no-var` rule for this file.

_The first commit should be reviewed with the `?w=1` flag for an easier diff._

timvandermeij committed:
Extract and modernize the webserver's request checking code

The `handler` method contained this code in two inline functions,
triggered via callbacks, which made the `handler` method big and harder
to read. Moreover, this code relied on variables from the outer scope,
which made it harder to reason about because the inputs and outputs
weren't easily visible.

This commit fixes the problems by extracting the request checking code
into a dedicated private method, and modernizing it to use e.g. `const`/
`let` instead of `var` and using template strings. The logic is now
self-contained in a single method that can be read from top to bottom
without callbacks and with comments annotating each check/section.

timvandermeij committed:
Modernize the webserver's handler method

This commit converts `var` to `const`/`let`, gives the variables more
readable names and annotates the code to make the flow clearer.

timvandermeij committed:
Modernize the remainder of the webserver's code and enable the `no-var` ESLint rule

This commit also moves the content type logic into a helper method to
ever so slightly reduce duplication.

timvandermeij commented:
/botio unittest

Snuffleupagus commented:
There's one more thing I'd suggest improving here, and that's the `WebServer` constructor; note how the parameters are set *after* initialization which seems a bit backwards:
 - https://github.com/mozilla/pdf.js/blob/f1a225889be50ee2d00d080449bce7f2f92a4d96/gulpfile.mjs#L2070-L2073
 - https://github.com/mozilla/pdf.js/blob/f1a225889be50ee2d00d080449bce7f2f92a4d96/test/test.mjs#L1018-L1023

It'd be much nicer if we directly passed in necessary parameters on initialization, by writing the constructor as follows:
```javascript
class WebServer {
  constructor({ root, host, port, cacheExpirationTime }) {
    this.root = root || ".";
    this.host = host || "localhost";
    this.port = port || 0;
    this.server = null;
    this.verbose = false;
    this.cacheExpirationTime = cacheExpirationTime || 0;
    this.disableRangeRequests = false;
    this.hooks = {
      GET: [crossOriginHandler],
      POST: [],
    };
  }
```

Snuffleupagus commented on test/webserver.mjs:
```diff
@@ -120,83 +117,83 @@ class WebServer {
-        checkFile
-      );
-      return;
+      pathPart = "test/resources/favicon.ico";
```
*I might be missing something, however:*
In the old code we'd *only* call the `checkFile` function and then return early here, but it appears that we're now (potentially) running a bunch more code in this case?

timvandermeij replied:
I was initially equally confused about the old code here, but it turns out the early return doesn't really achieve much. The `checkFile` call here in turn calls `statFile` which in turn serves the favicon file. However, note that the line below the early return at https://github.com/mozilla/pdf.js/blob/master/test/webserver.mjs#L133 does the exact same thing, just for a different path part, and the rest of the handler function below the early return doesn't actually do anything other than defining the inline functions and some variables. I also thought a lot more was happening here, but the inline functions made that hard to see.

Snuffleupagus replied:
Oh, that's right; I was getting confused by the "callback hell" of the old code!

This just shows *why* these kind of patches are so worthwhile, since the use of async/await means the code can now be read from top to bottom without any trouble!

Snuffleupagus commented on test/webserver.mjs:
```diff
@@ -120,83 +117,83 @@ class WebServer {
-        res.writeHead(500);
-        res.end();
+    // If a file is requested with range requests, serve it accordingly.
+    const range = request.headers.range;
```
Nit: `const { range } = request.headers;`

Snuffleupagus commented on test/webserver.mjs:
```diff
@@ -302,7 +297,7 @@ class WebServer {
     });
 
     const extension = path.extname(filePath).toLowerCase();
-    const contentType = mimeTypes[extension] || defaultMimeType;
+    const contentType = MIME_TYPES[extension] || DEFAULT_MIME_TYPE;
```
Given that we repeat this code multiple times, would a helper method make sense?
E.g. something like this:
```javascript
#getContentType(filePath) {
  const extension = path.extname(filePath).toLowerCase();
  return MIME_TYPES[extension] || DEFAULT_MIME_TYPE;
}

timvandermeij committed:
Improve the webserver's constructor

This makes the webserver configurable during instantiation rather than
having to set the parameters afterwards.

Snuffleupagus reviewed (approved):
r=me, thanks a lot for improving this code!

timvandermeij commented:
/botio unittest
